src/components/WalletConnect/NanoContract/NewNanoContractTransactionRequest.js

```diff
@@ -0,0 +1,317 @@
+ * @param {string} props.ncTxRequest.dapp.description
+ */
+export const NewNanoContractTransactionRequest = ({ ncTxRequest }) => {
+  const { nc, dapp } = ncTxRequest;
+  const dispatch = useDispatch();
+  const navigation = useNavigation();
+  const newTxStatus = useSelector((state) => state.walletConnect.newNanoContractTransaction.status);
+  const firstAddress = useSelector((state) => state.firstAddress);
+  // Nullable if the nano contract method is 'initialize'
+  const registeredNc = useSelector((state) => state.nanoContract.registered[nc.ncId]);
+  const knownTokens = useSelector((state) => ({ ...state.tokens, ...state.walletConnect.tokens }));
+  const blueprintInfo = useSelector((state) => state.nanoContract.blueprint[nc.blueprintId]);
+
+  const [showSelectAddressModal, setShowSelectAddressModal] = useState(false);
+  const [showDeclineModal, setShowDeclineModal] = useState(false);
+  /**
+   * If nano-contract's method is 'initialize' then the expression
+   * should be resolved to firstAddress value by default.
+   *
+   * In case of failure to load the first address the user will see
+   * a feedback message instruction it to select an address for the
+   * transaction.
+   */
+  const [ncAddress, setNcAddress] = useState(registeredNc?.address || firstAddress.address);
+  const ncToAccept = useMemo(() => ({
+    ...nc,
+    caller: ncAddress,
+  }), [ncAddress])
+
+  const toggleSelectAddressModal = () => setShowSelectAddressModal(!showSelectAddressModal);
+  const handleAddressSelection = (newAddress) => {
+    setNcAddress(newAddress);
+    toggleSelectAddressModal();
+  };
+
+  // Accepts the Nano Contract data preseted.
+  const onAcceptTransaction = () => {
+    // Update the caller with the address selected by the user.
+    const acceptedNc = { ...nc, caller: ncAddress };
+    // Signal the user has accepted the current request and pass the accepted data.
+    dispatch(walletConnectAccept(acceptedNc));
+  };
+
+  const onDeclineTransaction = () => {
+    setShowDeclineModal(true);
+  };
+  const onDeclineConfirmation = () => {
+    setShowDeclineModal(false);
+    dispatch(walletConnectReject());
+    navigation.goBack();
+  };
+  const onDismissDeclineModal = () => {
+    setShowDeclineModal(false);
+  };
+
+  // Control which content to show, if the nano contract is not registered
+  // a feedback content should tell user the nano contract must be registered first
+  // and only let user decline the transaction to get out the page, otherwise interaction
+  // content is showed.
+  const notInitialize = ncToAccept.method !== 'initialize';
+  const notRegistered = notInitialize && registeredNc == null;
+  // It results in true for registered nc and initialize request
+  const showRequest = !notRegistered;
+
+  // This effect should run only once because firstAddress is kept on state when loaded
+  useEffect(() => {
+    if (firstAddress.address) {
+      setNcAddress(firstAddress.address);
+    }
+  }, [firstAddress]);
```

Comment:
I didn't understand this `useEffect`. If it's a registered nano, shouldn't we have the registered address as the `ncAddress`?

Reply:
Yes, you right. I've added another condition for this assignment, the method should be `initialize`. Despite we have a the firstAddress as the default value for when the component is initializing, this state may not be initialized yet. This effect capture the change in the addressFirst when it is requested to load on `NanoContractExecInfo` child component.